pipelines/options.go

```diff
@@ -0,0 +1,309 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+)
+
+var (
+	// defaultSocketAPM specifies the socket path to use for connecting to the trace-agent.
+	// Replaced in tests
+	defaultSocketAPM = "/var/run/datadog/apm.socket"
+
+	// defaultSocketDSD specifies the socket path to use for connecting to the statsd server.
+	// Replaced in tests
+	defaultSocketDSD = "/var/run/datadog/dsd.socket"
+)
+
+// config holds the tracer configuration.
+type config struct {
+	// features holds the capabilities of the agent and determines some
+	// of the behaviour of the tracer.
+	features agentFeatures
+	// logToStdout reports whether we should log all traces to the standard
+	// output instead of using the agent. This is used in Lambda environments.
+	logToStdout bool
+	// logStartup, when true, causes various startup info to be written
+	// when the tracer starts.
+	logStartup bool
+	// service specifies the name of this application.
+	service string
+	// env contains the environment that this application will run under.
+	env string
+	// agentAddr specifies the hostname and port of the agent where the traces
+	// are sent to.
+	agentAddr string
+	// globalTags holds a set of tags that will be automatically applied to
+	// all spans.
+	globalTags map[string]interface{}
+	// httpClient specifies the HTTP client to be used by the agent's transport.
+	httpClient *http.Client
+	// hostname is automatically assigned when the DD_TRACE_REPORT_HOSTNAME is set to true,
+	// and is added as a special tag to the root span of traces.
+	hostname string
+	// dogstatsdAddr specifies the address to connect for sending metrics to the
+	// Datadog Agent. If not set, it defaults to "localhost:8125" or to the
+	// combination of the environment variables DD_AGENT_HOST and DD_DOGSTATSD_PORT.
+	dogstatsdAddr string
+	// statsd is used for tracking metrics associated with the runtime and the tracer.
+	statsd    statsd.ClientInterface
+	site      string
+	apiKey    string
+	agentLess bool
+}
+
+// StartOption represents a function that can be provided as a parameter to Start.
+type StartOption func(*config)
+
+// newConfig renders the tracer configuration based on defaults, environment variables
+// and passed user opts.
+func newConfig(opts ...StartOption) *config {
+	c := new(config)
+	c.agentAddr = defaultAddress
+	c.httpClient = defaultHTTPClient()
+	if v := os.Getenv("DD_ENV"); v != "" {
+		c.env = v
+	}
+	if v := os.Getenv("DD_SERVICE"); v != "" {
+		c.service = v
+	}
+	for _, fn := range opts {
+		fn(c)
+	}
+	if c.env == "" {
+		if v, ok := c.globalTags["env"]; ok {
+			if e, ok := v.(string); ok {
+				c.env = e
+			}
+		}
+	}
+	if c.service == "" {
+		if v, ok := c.globalTags["service"]; ok {
+			if s, ok := v.(string); ok {
+				c.service = s
+			}
+		} else {
+			c.service = filepath.Base(os.Args[0])
+		}
+	}
+	c.loadAgentFeatures()
+	if c.statsd == nil {
```

Comment:
Can we please explore a way to de-duplicate this code? Perhaps putting it in an `internal/agent` package. I think multiple packages would want to detect agent features, and equally will want to obtain a statsd client which is automatically set up to send to that agent.

The only problem that this creates is that a dependency on `dd-trace-go` is needed, which makes sub-modules a bit hard to maintain. But, as [my previous]() comment, I think we don't need a sub-module here.

Reply:
I will tackle this once we agree if we need a new sub-module or no.